### DISCUSSION TITLE
Pass initial values to Controller

### DISCUSSION THREAD

**Discussion by stevegcc** [ShadyBoukhary/flutter_clean_architecture#88] [Q&A] [answered]
I'm not sure if this section of the repo is monitored but I'll post here just in case.

I would like to create a widget that receives initial values of a list of items and an initial index through its constructor. The index would then update through a 'next' and 'previous' type of interaction. If I have understood this architecture correctly, I believe the Controller would be responsible for moving the index through the list of items and the widget would do something like `controller.item` to get the current item for display.

My question is, how do I get these initial values into the Controller?  I currently have the implementation below, where the `createState()` implementation appears to be frowned upon by the linter.

```
class WorkflowView extends View {
  final Set<Item> _items;
  final int _initialIndex;

  WorkflowView(
      {Key? key, required final Set<Item> items, required final int initialIndex})
      : _items = items,
        _initialIndex = initialIndex,
        super(key: key, routeObserver: AppRouter.getInstance().routeObserver);

  @override
  ViewState<WorkflowView, WorkflowController> createState() => _ExerciseWorkflowViewState(_initialIndex, _exercises);
}

class _WorkflowViewState extends ViewState<WorkflowView, WorkflowController> {
  _WorkflowViewState(final int initialIndex, final Set<Item> items)
      : super(WorkflowController(initialIndex, items));

  @override
  Widget get view {
    // Simplified, obviously 'next' and 'previous' buttons required etc.
    return Text(controller.item.name);
  }
}

```

**Comment by ShadyBoukhary** [comment]
Hi @stevegcc, this is the intended usage as shown in the example below:
https://github.com/ShadyBoukhary/flutter_clean_architecture/blob/master/example/lib/src/app/pages/home/home_view.dart#L14

What is the linter message exactly?

The return type in this line must match the type of the ViewState you're using. I think in your case it's _WorkflowViewState? Although you seem to be constructing a _ExerciseWorkflowViewState, which I cannot see in the example you posted.
```
 ViewState<WorkflowView, WorkflowController> createState() => _ExerciseWorkflowViewState(_initialIndex, _exercises);
```
Should be
```
  _ExerciseWorkflowViewState<WorkflowView, WorkflowController> createState() => _ExerciseWorkflowViewState(_initialIndex, _exercises);
```
Let me know if that resolves it for you.
In addition, you shouldn't keep a copy if the items and the index in the View. You should retrieve them from the controller. This won't cause issues however.

**Comment by stevegcc** [comment]
Thanks @ShadyBoukhary and apologies, I meant to include the linting error in the first post. This is the error I receive https://dart-lang.github.io/linter/lints/no_logic_in_create_state.html

I also made a mistake with the `createState` example in the code, I made the example names more concise and missed one. So it is the correct return type in reality.
`ViewState<WorkflowView, WorkflowController> createState() => _WorkflowViewState(_initialIndex, _exercises);`

I have the copy of the index and items in the view because I am displaying this view in a `showMaterialModalBottomSheet()` and passing the view in to the builder parameter as so 
`builder: WorkflowView(key: const Key('Workflow'), items: controller.items, initialIndex: index)`
From there, the values are passed into the State and then into the Controller. The Controller then has all logical functions for working with the items like `next`, `isFirst`, `getItem` and the view uses these.

What's the best method for getting these values into the controller? I am newish to Flutter and I don't seem to be able to find another way other than the source view passing the values into the new view via the constructor, which then needs to pass into the state and then the Controller. Do I need to put them into storage and retrieve them? 

This is possibly another question for another thread, however, before I refactored to use a BottomSheet, I was passing route parameters, extracting in the ViewState and setting the values in the Controller during the build method. This meant the items and index were mutable and I wasn't comfortable with that. Is there a way to access Route parameters in the Controller as well?

**Reply by ShadyBoukhary** [reply]
This linter message seems to be new. When i wrote the library that didn't exist. It seems like a stylistic choice the linter team made, as I don't see any drawbacks of using the current method you're using to pass default values into the controller. If you agree, you can disable that message by adding this comment to that line of code `// ignore: no_logic_in_create_state`.

Alternatively, there is another way to pass default values by overriding this method in your controller:
https://pub.dev/documentation/flutter_clean_architecture/latest/flutter_clean_architecture/Controller/onInitState.html

This method gets called during the ViewState `initState()` method. From there, you should be able to access the 
``` dart
final Set<Item> _items;
final int _initialIndex;
```

from your `WorkflowView` using this method
https://pub.dev/documentation/flutter_clean_architecture/latest/flutter_clean_architecture/Controller/getState.html

Example (simplified for clarity):
```dart
class MyController extends Controller {
      @override
      void onInitState() {
              this._initial_index = getState().widget._initial_index;
     }
}
```

Basically the Controller is accessing the ViewState, which has access to the View through the `widget` property.
https://api.flutter.dev/flutter/widgets/State/widget.html

This method would work but honestly if i were you i would disable that linter message because the 2nd method (which Flutter recommends) doesn't fit well with this architecture as it violates the rule that the controller should not access variables in the View. I'm not sure why the Flutter team decided this was bad practice, but there is no actual performance impact if you use the 1st method. I think it is purely a stylistic choice. So i would recommend the 1st method..

**Reply by ShadyBoukhary** [reply]
If you use the 2nd method, don't forget to add this line to your Build method:

https://github.com/ShadyBoukhary/flutter_clean_architecture/blob/master/example/lib/src/app/pages/home/home_view.dart#L28

**Reply by ShadyBoukhary** [reply]
Regarding your other question, can you open another discussion and add code samples to demonstrate what you are trying to do?
@stevegcc

**Reply by stevegcc** [reply]
Thanks @ShadyBoukhary  I have ended up with adding an ignore for the linting issue. It doesn't make much sense to me that we can't pass constructor arguments. It's not logic in my eyes and fulfils the "Implementations of createState() should return a new instance of a State object and do nothing more" statement the linter makes.

I will create some code examples for the other question and create a new question soon. I'm sure I'll run into that use case again. Thanks for your input on this.

**Comment by ShadyBoukhary** [comment, marked as answer]
Marking this comment as an answer
https://github.com/ShadyBoukhary/flutter_clean_architecture/discussions/88#discussioncomment-4595887
